core/src/main/scala/org/apache/spark/deploy/LocalSparkCluster.scala

```diff
@@ -63,23 +65,34 @@ class LocalSparkCluster(
 
     /* Start the Workers */
     for (workerNum <- 1 to numWorkers) {
-      val workerEnv = Worker.startRpcEnvAndEndpoint(localHostname, 0, 0, coresPerWorker,
-        memoryPerWorker, masters, null, Some(workerNum), _conf,
-        conf.get(config.Worker.SPARK_WORKER_RESOURCE_FILE))
+      val (workerEnv, workerRef) = Worker.startRpcEnvAndEndpoint(localHostname, 0, 0,
+        coresPerWorker, memoryPerWorker, masters, null, Some(workerNum), _conf,
+        conf.get(config.Worker.SPARK_WORKER_RESOURCE_FILE), isLocalCluster = true)
       workerRpcEnvs += workerEnv
+      workerRefs += workerRef
     }
 
     masters
   }
 
   def stop(): Unit = {
     logInfo("Shutting down local Spark cluster.")
+    // SPARK-31922: make sure all the workers have handled the messages(`KillExecutor`,
+    // `ApplicationFinished`) from the Master before we shutdown the workers' rpcEnvs.
+    // Otherwise, we could hit "RpcEnv already stopped" error.
+    var busyWorkers = workerRefs
+    while (busyWorkers.nonEmpty) {
+      Thread.sleep(300)
+      busyWorkers = busyWorkers.filterNot(_.askSync[Boolean](IsWorkerReadyToStop))
```

Comment:
Thank you for working on this, @Ngone51 . BTW, this seems to cause an infinite loop if a worker cannot response for some reasons. Can we use a finite `for` statement instead of `while` statement to avoid any potential issues?

Reply:
make sense, thanks!